### PR TITLE
dialects: add FSM dialect (WIP - refer to #1600)

### PR DESCRIPTION
**Pull request by luisacicolini**
Added the FSM dialect, currently working on file check testing.

### CODE REVIEW

**Comment by superlopuh** [comment]
could you please merge main into this branch? Also, could you please follow the steps in the PR template?

File: tests/filecheck/dialects/func/func_invalid.mlir [hunk 1]
```diff
@@ -31,3 +31,9 @@ builtin.module {
 }
 
 // CHECK: Expected all arguments to be named or all arguments to be unnamed.
+
+// -----
+
+"func.call"() { "callee" = @call::@invalid } : () -> ()
```
Comment (compor) [inline tests/filecheck/dialects/func/func_invalid.mlir:37]: This and other parts of this PR seem to have crept up from https://github.com/xdslproject/xdsl/pull/1598.
Could you please separate the two?

**Review by compor — CHANGES_REQUESTED** [review]

File: tests/filecheck/dialects/func/func_invalid.mlir (hunk 1 above)
Comment (luisacicolini) [reply tests/filecheck/dialects/func/func_invalid.mlir:37]: Yes so sorry I messed up with some branch, gonna solve that now!

File: xdsl/dialects/fsm.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+
+
+@irdl_op_definition
+class Return(IRDLOperation):
+    """marks the end of a region of `fsm.transition` and return
+    values if the parent region is a `$guard` region"""
+
+    name = "fsm.return"
+
+    operand = operand_def(signlessIntegerLike)
```
Comment (compor) [inline xdsl/dialects/fsm.py:242]: I think you need:

```suggestion
    operand = opt_operand_def(signlessIntegerLike)
```

File: xdsl/dialects/fsm.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+
+    def verify_(self) -> None:
+        if not self.operands:
+            return
+
+        parent = self.parent_op()
+        if not isinstance(parent, Transition):
+            return
+
+        if self.parent_region() != parent.guard:
```
Comment (compor) [inline xdsl/dialects/fsm.py:260]: I'd make the conditions explicit instead of separating the tests:

```suggestion
        if self.operands and ...:
```

**Review by compor — CHANGES_REQUESTED** [review]

File: xdsl/dialects/fsm.py (hunk 3 above)
Comment (luisacicolini) [reply xdsl/dialects/fsm.py]: Thanks a lot for all your observations! :)

Comment (luisacicolini) [reply xdsl/dialects/fsm.py]: Concerning this one I actually thought more in detail about it and probably there is no need to raise an exception

File: tests/filecheck/dialects/func/func_invalid.mlir [hunk 4]
```diff
@@ -30,4 +30,4 @@ builtin.module {
   }
 }
 
-// CHECK: Expected all arguments to be named or all arguments to be unnamed.
+// CHECK: Expected all arguments to be named or all arguments to be unnamed.
```
Comment (tobiasgrosser) [inline tests/filecheck/dialects/func/func_invalid.mlir:33]: This change is unintentional, no?

**Review by tobiasgrosser — COMMENTED** [review]
Nice. I guess we will need some more FileCheck test cases.

File: tests/filecheck/dialects/func/func_invalid.mlir (hunk 4 above)
Comment (luisacicolini) [reply tests/filecheck/dialects/func/func_invalid.mlir:33]: Indeed! I shall try andrevert it today

**Review by webmiche — COMMENTED** [review]
This is a mirrored dialect from the CIRCT FSM one, right? Please add documentation with a link to that dialect.

Do you guys think it would make sense to implement differential testing here? That would mean we would have to build the circt-based opt tool on the CI, which might take quite a bit longer, but it would give us more certainty on the structure of this dialect and that interoperation works properly...

**Comment by tobiasgrosser** [comment]
I feel CIRCT testing is a little much. We are just trying to get @luisacicolini started on xDSL. Let's not make the initial price of using and playing with it too high.

**Comment by luisacicolini** [comment]
> This is a mirrored dialect from the CIRCT FSM one, right? Please add documentation with a link to that dialect.
> 
> Do you guys think it would make sense to implement differential testing here? That would mean we would have to build the circt-based opt tool on the CI, which might take quite a bit longer, but it would give us more certainty on the structure of this dialect and that interoperation works properly...

It is! I will add the doc, thank you

File: xdsl/dialects/fsm.py [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+
+@irdl_op_definition
+class Output(IRDLOperation):
+    """Represents the outputs of a machine under a specific state. The
+    types of `$operands` should be consistent with the output types of the state
+    machine"""
+
+    name = "fsm.output"
+
+    operand = var_operand_def(AnyAttr())
```
Comment (superlopuh) [inline xdsl/dialects/fsm.py:220]:
```suggestion
    operand = var_operand_def()
```

Default is unconstrained

File: xdsl/dialects/fsm.py [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+        operand: Sequence[SSAValue | Operation],
+    ):
+        super().__init__(
+            operands=[operand],
+        )
+
+    def verify_(self):
+        parent = self.parent_op()
+        res_type_machine = type(None)
+        while HasParent(Machine):
```
Comment (superlopuh) [inline xdsl/dialects/fsm.py:233]: This seems off, it does not seem to reference anything in the local context.

File: xdsl/dialects/fsm.py [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+            operands=[variable, value],
+        )
+
+    def verify_(self):
+        if not (SymbolOpInterface(self.variable) == Variable):
+            raise VerifyException(
+                "The definition operator of variable attribute must be Vatiable"
+            )
+
+        if not (HasParent(Transition)):
```
Comment (superlopuh) [inline xdsl/dialects/fsm.py:399]: same here this seems to not reference anything in the local context

**Review by superlopuh — CHANGES_REQUESTED** [review]
Thank you for cleaning up the changes! After thinking about this a little longer, it does feel like adding these one op at a time might be easier. Or at least potentially adding a single op to start with, and then PRs for more as we go. Otherwise it's quite difficult to keep track of all the different changes across the ops in this dialect. Hopefully it won't be too difficult to split out one first op to add, and then see about the rest.

File: tests/filecheck/dialects/fsm/fsm_op.mlir [hunk 8]
```diff
@@ -1,15 +1,89 @@
...
+    }) {sym_name = "B"} : () -> ()
+    "fsm.state"() ({
+      "fsm.output"(%arg0) : (i1) -> ()
+    }, {
+      "fsm.transition"() ({
+      }, {
+      }) {nextState = @C} : () -> ()
+    }) {sym_name = "C"} : () -> ()
+  }) {function_type = (i1) -> i1, initialState = "A", sym_name = "foo"} : () -> ()
+}) : () -> ()
```
Comment (superlopuh) [inline tests/filecheck/dialects/fsm/fsm_op.mlir:89]:
```suggestion
}) : () -> ()

```

File: xdsl/dialects/fsm.py [hunk 9]
```diff
@@ -21,6 +21,7 @@
     SSAValue,
     TypeAttribute,
 )
+from xdsl.ir.core import Dialect
```
Comment (superlopuh) [inline xdsl/dialects/fsm.py:24]:
```suggestion
    TypeAttribute,
    Dialect,
)
```

**Comment by superlopuh** [comment]
Ok sounds good, thank you

File: xdsl/dialects/fsm.py (hunk 3 above)
Comment (compor) [reply xdsl/dialects/fsm.py]: That is fine!

**Comment by luisacicolini** [comment]
Please refer to #1600 as it contains the first batch of operators. As those are solid and merged in the main I will proceed with the more complex ones (HWinstance, Instance, Trigger). Thank you all for your suggestions :)

**Comment by compor** [comment]
@luisacicolini I think we should close this PR since work is either partly in or being worked on the other PRs. WDYT?

**Pull request closed without merging**